ddtrace/tracer/handler.go

```diff
@@ -0,0 +1,167 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package tracer
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/log"
+)
+
+type traceHandler interface {
+	// pushPayload adds a trace to the current payload being constructed by the handler.
+	pushPayload([]*span)
+	// flush should cause the handler to send its current payload. The flush can happen asynchronously, but
+	// the handler must be ready to accept new traces with pushPayload when flush() returns.
+	flush()
+	// wait will pause until the traceHandler is finished any asynchronous tasks it is performing.
+	wait()
+	// sample gives a handler the opportunity to perform sampling on a span. This is called when a span
+	// is created.
+	sample(*span)
```

Comment:
Is there any way we can remove `sample` and `wait` from the interface?

Reply:
I think we can get rid of `wait` if we let the handlers use the tracer's `WaitGroup`. I was hoping to eliminate as much shared state as possible between the tracer and the handlers, but sharing the `WaitGroup` is probably ok.

We might be able to get rid of `sample`, but I'm not sure there's a clean way. The `tracer` object needs to call `sample` and the `agentTraceHandler` object needs to receive feedback from the agent, so either they both need a reference to a shared `prioritySampler`, or there needs to be some sort of abstraction like this. Unlike the above, where sharing the `WaitGroup` is acceptable, I don't really like the idea of sharing the `prioritySampler` and having the `agentHandler` messing around with state inside the `tracer`. It's only applicable when sending to the agent and the `lambdaHandler` won't use it, so it seems like it belongs with the `agentHandler`.

Let me know if you have ideas about this.

Reply:
Here's the way I'd do it (not to say you should do the same):

* Aim to keep concerns fully separated and interface as slim as possible
* Call the interface `traceWriter` or `payloadSender`
* Rename `pushPayload` to `add`, we're not pushing a payload, we're adding a trace to it
* Remove `sample` and simply do something like:
```go
if !t.config.lambda {
    // non-lambda traces need priority sampling
    t.prioritySampling.apply(span)
}
```
It's not ideal, but arguably cleaner and worth it?
* Remove `wait`
* Remove the ticker from the tracer and have each handler/writer have their own ticker
* Repurpose `flush` so that when it's called, it flushes whatever's left in the handler/writer and waits for it to finish. Basically `flush` would become what is now `flush` + `wait`, a shutdown mechanism. Maybe better called `stop`

Then the interface would be:
```go
type traceWriter interface {
    // add adds the trace to the traceWriter to be written to its destination.
    add(trace []*span)

    // stop stops the traceWriter and flushes any remaining spans.
    stop() 
}
```
The constructor for each implementation would start a ticker goroutine.

Reply:
I will try something like you suggest.

The priority sampling solution of checking for lambda will not be sufficient.
```go
if !t.config.lambda {
    // non-lambda traces need priority sampling
    t.prioritySampling.apply(span)
}
```

The issue is that the agent `traceWriter` needs to submit agent responses into the `prioritySampler` (https://github.com/DataDog/dd-trace-go/blob/v1/ddtrace/tracer/tracer.go#L361)

I initially had a ticker per `traceWriter` like you suggested, but this involved duplicating the `worker` function, which I wanted to avoid. I'll put up that change in and we can see if the tradeoff (more duplication for a simpler interface) is worth it.